app/models/user_annotation.rb

```diff
@@ -0,0 +1,291 @@
+  #cluster is current cluster
+  def create_array(cluster, threshold, annotation, user_data_arrays_attributes)
+    #create a hash of cell names of format
+    #{cell_name1=>'its label', cell_name_2=>'its label'}
+    user_annotation_map = {}
+    user_data_arrays_attributes.keys.each do |key|
+      name = user_data_arrays_attributes[key][:name]
+      user_data_arrays_attributes[key][:values].split(',').each do |cell_name|
+        user_annotation_map[cell_name] = name
+      end
+    end
+
+    #get the cell names, x values and y values this selection was created on
+    cell_name_array = cluster.concatenate_data_arrays('text', 'cells', threshold, annotation)
+    x_array = cluster.concatenate_data_arrays('x', 'coordinates', threshold, annotation)
+    y_array = cluster.concatenate_data_arrays('y', 'coordinates', threshold, annotation)
+
+    #user data arrays are different than normal data arrays, so if you created a selection from a user selection this:
+    #gets the cell names, x values and y values this selection was created on
+    if !annotation.nil? and annotation.include? 'user'
+      user_annot = UserAnnotation.where(name: annotation.gsub('--group--user','')).first
+      cell_name_array = user_annot.concatenate_user_data_arrays('text', 'cells', threshold, annotation)
+      x_array = user_annot.concatenate_user_data_arrays('x', 'coordinates', threshold, annotation)
+      y_array = user_annot.concatenate_user_data_arrays('y', 'coordinates', threshold, annotation)
+    end
+
+    #get an ordered array of labels
+    annotation_array = []
+    #check if any of the values are undefined, aka this data array is being extrapolated
+    undefined_happened = false
+
+    cell_name_array.each do |name|
+      cell_name = user_annotation_map[name]
+      if cell_name.nil?
+        cell_name = 'Undefined'
+        undefined_happened = true
+      end
+      annotation_array << cell_name
+    end
+
+
+    #Create the name of the subsample annotation
+    sub_an = self.name + '--group--user'
+
+    #Slice the arrays if into segments smaller than the maximum size (100k)
+    x_arrays = x_array.each_slice(UserDataArray::MAX_ENTRIES).to_a
+    y_arrays = y_array.each_slice(UserDataArray::MAX_ENTRIES).to_a
+    cell_name_arrays = cell_name_array.each_slice(UserDataArray::MAX_ENTRIES).to_a
+
+    #create each data array
+    annotation_array.each_slice(UserDataArray::MAX_ENTRIES).each_with_index do |val, i|
+      #if threshold exists then the subsample annotation and threshold need to be set on the created data array
+      if !threshold.nil?
+        #Create annotation array
+        UserDataArray.create(name: name, array_type: 'annotations', values: val, cluster_name: cluster.name, array_index: i+1, subsample_threshold: threshold, subsample_annotation: sub_an, user_id: self.user_id, cluster_group_id: self.cluster_group_id, study_id: self.study_id, user_annotation_id: id)
+        logger.info('created annotation')
```

Comment:
These logging statements are uninformative - you should always have a timestamp with Time.now and say with much more detail what you are doing.  This will be needed if we are trying to debug issues in production.  Look at the types of logging statements in the Study class when we are parsing uploaded files to get an idea of how to reformat these.